app/gui/src/controller/ide.rs

```diff
@@ -150,18 +157,44 @@ pub trait ManagingProjectAPI {
     /// and then for the project opening.
     fn open_project_by_name(&self, name: String) -> BoxFuture<FallibleResult> {
         async move {
-            let projects = self.list_projects().await?;
-            let mut projects = projects.into_iter();
-            let project = projects.find(|project| project.name.as_ref() == name);
-            let uuid = project.map(|project| project.id);
-            if let Some(uuid) = uuid {
-                self.open_project(uuid).await
-            } else {
-                Err(ProjectNotFound(name).into())
+            let project_id = self.find_project(&ProjectToOpen::Name(name.into())).await?;
+            self.open_project(project_id).await
+        }
+        .boxed_local()
+    }
+
+    /// Open a project by name or ID. If no project with the given name exists, it will be created.
+    fn open_or_create_project(&self, project_to_open: ProjectToOpen) -> BoxFuture<FallibleResult> {
+        async move {
+            match self.find_project(&project_to_open).await {
+                Ok(project_id) => self.open_project(project_id).await,
+                Err(error) =>
+                    if let ProjectToOpen::Name(name) = project_to_open {
+                        info!("Attempting to create project with name {}", name);
+                        self.create_new_project(Some(name.to_string()), None).await
+                    } else {
+                        Err(error)
+                    },
             }
         }
         .boxed_local()
     }
+
+    /// Find a project by name or ID.
+    fn find_project<'a, 'b: 'a>(
+        &'a self,
+        project_to_open: &'b ProjectToOpen,
```

Comment:
Why do we need so complex lifetime dependencies? I do not see in this code anything that would require it – it's very rare that something like that is needed.

Reply:
The returned `BoxFuture` captures `'a` "by default", so `'b: 'a` ensures that it can also safely capture `project_to_open`. But to make it more clear, I'd go with this signature:
```rust
    /// Find a project by name or ID.
    fn find_project<'a, 'b, 'c: 'a + 'b>(
        &'a self,
        project_to_open: &'b ProjectToOpen,
    ) -> BoxFuture<'c, FallibleResult<Uuid>>
```
(or `'a: 'c, 'b: 'c, 'c` - I don't remember how the lifetime bounds exactly work).

Reply:
Thanks for explaining it, @farmaazon. I agree that the complexer signature makes it even more clear. I'll change it to that.

(It's indeed `'a: 'c, 'b: 'c, 'c`: you can remember it by reading `:` as "outlives".)